File: modules/fbs-core/api/src/main/kotlin/de/thm/ii/fbs/controller/v2/CourseRegistrationController.kt
package de.thm.ii.fbs.controller.v2

import com.fasterxml.jackson.databind.JsonNode
import de.thm.ii.fbs.model.v2.Course
import de.thm.ii.fbs.model.v2.CourseRole
import de.thm.ii.fbs.model.v2.GlobalRole
import de.thm.ii.fbs.model.v2.Participant
import de.thm.ii.fbs.model.v2.security.LegacyToken
import de.thm.ii.fbs.services.v2.persistence.CourseRegistrationRepository
import de.thm.ii.fbs.utils.v2.annotations.CurrentToken
import de.thm.ii.fbs.utils.v2.exceptions.ForbiddenException
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(path = ["/api/v1/"], produces = [MediaType.APPLICATION_JSON_VALUE])
class CourseRegistrationController(
    private val courseRegistrationRepository: CourseRegistrationRepository
) {
    /**
     * Get registered courses
     * @param uid User id
     * @return List of courses
     */
    @GetMapping("/users/{uid}/courses")
    @ResponseBody
    fun getRegisteredCourses(@CurrentToken currentToken: LegacyToken, @PathVariable("uid") uid: Int): List<Course> {
        val globalRole = currentToken.globalRole

        if (globalRole == GlobalRole.ADMIN || globalRole == GlobalRole.MODERATOR || currentToken.id == uid) {
            return courseRegistrationRepository.findAllCoursesByUid(uid)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Get participants of a course
     * @param cid Course id
     * @return List of courses
     */
    @GetMapping("/courses/{cid}/participants")
    @ResponseBody
    fun getParticipants(@CurrentToken currentToken: LegacyToken, @PathVariable("cid") cid: Int): List<Participant> {
        val globalRole = currentToken.globalRole
        val participants = courseRegistrationRepository.findAllParticipantsByCourseId(cid)

        val privilegedByCourse: CourseRole = participants.find { participant: Participant -> participant.user.id == currentToken.id }?.role
            ?: CourseRole.STUDENT

        if (privilegedByCourse != CourseRole.STUDENT || globalRole != GlobalRole.USER) {
            return participants
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Register a user into a course
     * @param uid User id
     * @param cid Course id
     * @param body Content
     */
    @PutMapping("/users/{uid}/courses/{cid}", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun register(@CurrentToken currentToken: LegacyToken, @PathVariable("uid") uid: Int, @PathVariable("cid") cid: Int, @RequestBody body: JsonNode) {
        val role: CourseRole = CourseRole.valueOf(body["roleName"].asText())
        val privileged = (currentToken.globalRole == GlobalRole.ADMIN || currentToken.globalRole == GlobalRole.MODERATOR) ||
            courseRegistrationRepository.findAllCourseRolesOfCourseIdsByUser(currentToken.id)
            .getOrElse(cid) { CourseRole.STUDENT } == CourseRole.DOCENT

        if (privileged) {
            courseRegistrationRepository.register(cid, uid, role.ordinal)
        } else if (uid == currentToken.id) {
            courseRegistrationRepository.register(cid, uid, CourseRole.STUDENT.ordinal)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Deregister a user from a course
     * @param uid Course id
     * @param cid Course id
     */
    @DeleteMapping("/users/{uid}/courses/{cid}")
    fun deregister(@CurrentToken currentToken: LegacyToken, @PathVariable("uid") uid: Int, @PathVariable("cid") cid: Int) {
        val privileged = currentToken.globalRole.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
            courseRegistrationRepository.findAllCourseRolesOfCourseIdsByUser(currentToken.id)
            .getOrElse(cid) { CourseRole.STUDENT } == CourseRole.DOCENT

        if (privileged || uid == currentToken.id) {
            courseRegistrationRepository.deleteByCourseIdAndUserId(cid, uid)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Deregister all users with a specific role from a course
     * @param cid Course id
     * @param body Content
     */
    @PutMapping("/courses/{cid}/deregisterrole", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun deregisterRole(@CurrentToken currentToken: LegacyToken, @PathVariable("cid") cid: Int, @RequestBody body: JsonNode) {
        val role: CourseRole = CourseRole.valueOf(body["roleName"].asText())

        val privileged = currentToken.globalRole.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
            courseRegistrationRepository.findAllCourseRolesOfCourseIdsByUser(currentToken.id)
            .getOrElse(cid) { CourseRole.STUDENT } == CourseRole.DOCENT

        if (privileged) {
            courseRegistrationRepository.deleteCourseRoleByCidAndCourseRole(cid, role.ordinal)
        } else {
            throw ForbiddenException()
        }
    }

    /**
     * Deregister all user except the current user
     * @param cid Course id
     */
    @GetMapping("/courses/{cid}/deregisterall")
    fun deregisterAll(@CurrentToken currentToken: LegacyToken, @PathVariable("cid") cid: Int) {
        val privileged = currentToken.globalRole.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
            courseRegistrationRepository.findAllCourseRolesOfCourseIdsByUser(currentToken.id)
            .getOrElse(cid) { CourseRole.STUDENT } == CourseRole.DOCENT

        if (privileged) {
            courseRegistrationRepository.deleteAllUsersByCourseId(cid, currentToken.id)
        } else {
            throw ForbiddenException()
        }
    }
}

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/CourseRegistrationController.scala
package de.thm.ii.fbs.controller

import com.fasterxml.jackson.databind.JsonNode
import de.thm.ii.fbs.controller.exception.ForbiddenException
import de.thm.ii.fbs.model.{Course, CourseRole, GlobalRole, Participant}
import de.thm.ii.fbs.services.persistence.{CourseRegistrationService, CourseService}
import de.thm.ii.fbs.services.security.AuthService
import de.thm.ii.fbs.util.JsonWrapper.jsonNodeToWrapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation._

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}

/**
  * Handles course registration and course participants
  */
@RestController
@CrossOrigin
@RequestMapping(path = Array("/api/v1/"), produces = Array(MediaType.APPLICATION_JSON_VALUE))
class CourseRegistrationController {
  @Autowired
  private val authService: AuthService = null
  @Autowired
  private val courseRegistrationService: CourseRegistrationService = null
  @Autowired
  private val courseService: CourseService = null

  /**
    * Get registered courses
    *
    * @param uid User id
    * @param req http request
    * @param res http response
    * @return List of courses
    */
  @GetMapping(value = Array("/users/{uid}/courses"))
  @ResponseBody
  def getRegisteredCourses(@PathVariable("uid") uid: Integer, req: HttpServletRequest, res: HttpServletResponse): List[Course] = {
    val user = authService.authorize(req, res)
    val globalRole = user.globalRole

    if (globalRole == GlobalRole.ADMIN || globalRole == GlobalRole.MODERATOR || user.id == uid) {
      courseRegistrationService.getRegisteredCourses(uid, ignoreHidden = false)
    } else {
      throw new ForbiddenException()
    }
  }

  /**
    * Get participants of a course
    *
    * @param cid Course id
    * @param req http request
    * @param res http response
    * @return List of courses
    */
  @GetMapping(value = Array("/courses/{cid}/participants"))
  @ResponseBody
  def getParticipants(@PathVariable("cid") cid: Integer, req: HttpServletRequest, res: HttpServletResponse): List[Participant] = {
    val user = authService.authorize(req, res)
    val globalRole = user.globalRole
    val participants = courseRegistrationService.getParticipants(cid)

    val privilegedByCourse = participants.find(_.user.id == user.id).map(_.role).exists {
      case CourseRole.DOCENT | CourseRole.TUTOR => true
      case _ => false
    }

    (globalRole, privilegedByCourse) match {
      case (GlobalRole.ADMIN | GlobalRole.MODERATOR, _) | (_, true) => participants
      case _ => throw new ForbiddenException()
    }
  }

  /**
    * Register a user into a course
    *
    * @param uid  User id
    * @param cid  Course id
    * @param req  http request
    * @param res  http response
    * @param body Content
    */
  @PutMapping(value = Array("/users/{uid}/courses/{cid}"), consumes = Array(MediaType.APPLICATION_JSON_VALUE))
  def register(@PathVariable("uid") uid: Int, @PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse,
               @RequestBody body: JsonNode): Unit = {
    val user = authService.authorize(req, res)
    val role = Option(body).flatMap(_.retrive("roleName").asText()).map(CourseRole.parse).getOrElse(CourseRole.STUDENT)
    val password = body.retrive("password").asText()

    val course = courseService.findByPassword(cid, password)
    course match {
      case Some(_) =>
        val privileged = user.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
          courseRegistrationService.getCoursePrivileges(user.id).getOrElse(cid, CourseRole.STUDENT) == CourseRole.DOCENT

        (privileged, user.id) match {
          case (true, _) => courseRegistrationService.register(cid, uid, role)
          case (false, `uid`) => courseRegistrationService.register(cid, uid, CourseRole.STUDENT)
          case _ => throw new ForbiddenException()
        }
      case None => throw new ForbiddenException()
    }
  }

  /**
    * Deregister a user from a course
    *
    * @param uid Course id
    * @param cid Course id
    * @param req http request
    * @param res http response
    */
  @DeleteMapping(value = Array("/users/{uid}/courses/{cid}"))
  def deregister(@PathVariable("uid") uid: Int, @PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse): Unit = {
    val user = authService.authorize(req, res)

    val privileged = user.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
      courseRegistrationService.getCoursePrivileges(user.id).getOrElse(cid, CourseRole.STUDENT) == CourseRole.DOCENT

    (privileged, user.id) match {
      case (true, _) | (_, `uid`) => courseRegistrationService.deregister(cid, uid)
      case _ => throw new ForbiddenException()
    }
  }

  /**
    * Deregister all users with a specific role from a course
    *
    * @param cid  Course id
    * @param req  http request
    * @param res  http response
    * @param body Content
    */
  @PutMapping(value = Array("/courses/{cid}/deregisterrole"), consumes = Array(MediaType.APPLICATION_JSON_VALUE))
  def deregisterRole(@PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse,
                     @RequestBody body: JsonNode): Unit = {
    val user = authService.authorize(req, res)
    val role = Option(body).flatMap(_.retrive("roleName").asText()).map(CourseRole.parse).getOrElse(CourseRole.STUDENT)

    val privileged = user.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
      courseRegistrationService.getCoursePrivileges(user.id).getOrElse(cid, CourseRole.STUDENT) == CourseRole.DOCENT

    (privileged, user.id) match {
      case (true, _) => courseRegistrationService.deregisterRole(cid, role)
      case _ => throw new ForbiddenException()
    }
  }

  /**
    * Deregister all user except the current user
    *
    * @param cid Course id
    * @param req http request
    * @param res http response
    */
  @GetMapping(value = Array("/courses/{cid}/deregisterall"))
  def deregisterAll(@PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse): Unit = {
    val user = authService.authorize(req, res)
    val uid = user.id

    val privileged = user.hasRole(GlobalRole.ADMIN, GlobalRole.MODERATOR) ||
      courseRegistrationService.getCoursePrivileges(user.id).getOrElse(cid, CourseRole.STUDENT) == CourseRole.DOCENT

    (privileged, user.id) match {
      case (true, _) => courseRegistrationService.deregisterAll(cid, uid)
      case _ => throw new ForbiddenException()
    }
  }
}
